rfcs/20250428-graph-api-sdpa-training/README.md

```diff
@@ -0,0 +1,620 @@
+| **grad_attn_bias**    |                     |                     | ✓                       | ✓                | Gradient of the attention mask tensor (if applicable).                         |
+
+## Proposals
+
+The following proposals outline different approaches to support SDPA training
+features. Each proposal requires a decision.
+
+Based on the investigation, the most frequently used features for SDPA training
+are dropout and softmax statistics (`log_sumexp`). Below are the proposals to
+support these features.
+
+### Proposal 1: Support Dropout
+
+#### Proposal 1.A: No Support for Dropout
+
+In this proposal, the oneDNN Graph API will not support the Dropout operation.
+
+##### Pros
+
+- No changes are required to the oneDNN Graph API.
+- The training forward graph remains similar to the inference graph.
+
+##### Cons
+
+- Models using Dropout in SDPA training cannot rely on oneDNN Graph for computation.
+
+#### Proposal 1.B: Add Dropout Operation to oneDNN Graph API
+
+The proposed definition for the Dropout operation is as follows:
+
+| **Inputs/Attributes/Outputs** | **Name**           | **Description**                                                               |
+|-------------------------------|--------------------|-------------------------------------------------------------------------------|
+| Input                         | `src`             | The tensor to which Dropout will be applied.                                   |
+| Input                         | `seed`            | Seed for the Philox random number generator used in Dropout.                   |
+| Input                         | `offset`          | Offset for the Philox random number generator used in Dropout.                 |
+| Attribute                     | `p`               | The dropout probability (e.g., 0.1 for 10% dropout), with values in the range [0, 1). |
+| Output                        | `dst`             | The input tensor with randomly dropped elements scaled by the inverse of the keep probability (`1 - p`). |
+
+The backward pass for Dropout is mathematically identical to the forward pass.
+The only difference lies in the interpretation of the input (`src` in the
+forward pass vs. `diff_dst` in the backward pass) and the output (`dst` in the
+forward pass vs. `diff_src` in the backward pass). Therefore, a dedicated
+`DropoutBackward` operation is not required.
+
+##### Pros
+
+- Enables users to incorporate Dropout in their models.
+- Bridges the gap between the oneDNN Graph API and the PyTorch SDPA API.
+
+##### Cons
+
+- Requires adding a new Dropout operation to the oneDNN Graph API.
+- A reference implementation for Dropout is needed. Currently, the oneDNN
+  Primitive API has limited support for Dropout, implemented as a primitive
+  attribute for `MatMul` without Philox offset input. Supporting SDPA training
+  may require extending this implementation.
+
+#### Proposal 1.C: Add RNG (Random Number Generator) Operation to oneDNN Graph API
```

Comment:
For CPU, will the API be different if we use another generator other than RNG?

Reply:
IIUC, you mean `will the API be different if you use another generator other than Philox`. 

I think there will be at least one more attribute for selecting different generator algorithms, e.g. Philox, MT19937. And if different generator requires different inputs (e.g. seed and offset for Philox, and maybe other inputs for MT19937), then other inputs may be added.

So the question will be what `another generator` will you use and what inputs are needed for it?

Reply:
Yes, you are right that I refer to `Philox` here. As CPU has not the request yet, I leave this as an open. Just to confirm if other generators may impact the API.